sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs

```diff
@@ -188,16 +222,17 @@ private bool HasSafeBinderInitialization(InitializerExpressionSyntax initializer
                     return false;
                 }
 
-                var typeSymbol = semanticModel.GetTypeInfo(binderAssignment.Right).Type;
-                return typeSymbol != null && IsBinderSafe(typeSymbol);
+                return semanticModel.GetTypeInfo(binderAssignment.Right).Type is {} typeSymbol &&
+                       IsTypeSafe(typeSymbol);
             }
 
-            private bool IsBinderSafe(ITypeSymbol symbol) =>
-                binderValidityMap.GetOrAdd(symbol, typeSymbol =>
+            private bool IsTypeSafe(ITypeSymbol symbol) =>
+                symbolValidityMap.GetOrAdd(symbol, typeSymbol =>
                 {
-                    var declaration = GetBindToTypeMethodDeclaration(typeSymbol);
+                    var declaration = GetBindToTypeMethodDeclaration(typeSymbol) ??
+                                      GetResolveTypeMethodDeclaration(typeSymbol);
```

Comment:
Although unlikely, if we have a `JavaScriptTypeResolver` that implements `BindToType`, its the `BindToType` implementation that will be checked here instead of the `ResolveType`.
I think it's ok as the likelihood of this actually happening is very small, but it can be worth it to document as a unit test

Reply:
I've added a fix and some test cases for this.